### PR TITLE
Version 1.1.0

### PR DESCRIPTION
**Pull request by ghbm-itk**
### Changed

- Orchestrator rewritten to use NiceGui instead of TKinter for its UI.
- Lots and lots of UI changes and upgrades.
- All UI logic removed from db_util.
- Refactoring and renaming of files.
- Scheduler log area limited to 1000 lines.
- Updated Github Workflows

### Added

- Queues tab in Orchestrator to monitor queues.
- Scheduler logs errors when a robot returns a non-zero return code.
- Ability to change the next run time on Scheduled triggers.
- Automated tests of database functionality and OrchestratorConnection.
- Orchestrator auto updates data when in focus.

### Fixed

- Scheduler doesn't spam "File not found" anymore.
- Type hints on db functions.

### CODE REVIEW

**Comment by ghbm-itk** [comment]
Linting workflow passed on forked repo.
https://github.com/ghbm-itk/OpenOrchestrator/actions

File: .github/workflows/Linting.yml [hunk 1]
```diff
@@ -1,24 +1,32 @@
-name: Pylint
+name: Linting
 
-on: [pull_request]
+on: [push]
```
Comment (rimi-itk) [inline .github/workflows/Linting.yml:3]: This does not seem right. All code changes are made via pull requests and there the task should run `on: [pull_request]`.

File: OpenOrchestrator/database/queues.py [hunk 2]
```diff
@@ -38,6 +39,20 @@ class QueueElement(Base):
...
 
+    def to_row_dict(self) -> dict:
+        """Convert the object to a dict for display in a table."""
+        return {
+            "ID": self.id,
+            "Reference": self.reference,
+            "Status": self.status,
+            "Data": self.data,
+            "Created Date": self.created_date.strftime("%d-%m-%Y %H:%M:%S"),
+            "Start Date": self.start_date.strftime("%d-%m-%Y %H:%M:%S") if self.start_date else 'N/A',
```
Comment (rimi-itk) [inline OpenOrchestrator/database/queues.py:50]: 🎊

File: OpenOrchestrator/orchestrator/application.py [hunk 3]
```diff
@@ -1,41 +1,77 @@
...
+            self.c_tab = ConstantTab("Constants")
+            self.q_tab = QueueTab("Queues")
+            SettingsTab('Settings')
+
+        self._define_on_close()
+
+        app.on_connect(self.update_loop)
+        app.on_disconnect(app.shutdown)
+        app.on_exception(lambda exc: ui.notify(exc, type='negative'))
+        ui.run(title="Orchestrator", favicon='🤖', native=False, port=23406, reload=False)
```
Comment (rimi-itk) [inline OpenOrchestrator/orchestrator/application.py:42]: I prefer 🍰 .

File: OpenOrchestrator/orchestrator/popups/trigger_popup.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        self._define_validation()
+
+        if trigger:
+            self._pre_populate()
+
+    def _define_validation(self):
+        self.trigger_input.validation = {"Please enter a trigger name.": bool}
+        self.name_input.validation = {"Please enter a process name.": bool}
+        self.path_input.validation = {"Please enter a process path.": bool}
+        self.queue_input.validation = {"Please enter a queue name.": bool}
```
Comment (rimi-itk) [inline OpenOrchestrator/orchestrator/popups/trigger_popup.py:76]: This is inconsistent with https://github.com/itk-dev-rpa/OpenOrchestrator/pull/58/files#diff-645d151cfc255c2311a8f1fcb9529b360f7db2fcda57273e88943dd292965675R52-R54 where you don't end validation messages with a period. Validation messages should be consistently formatted.

File: OpenOrchestrator/orchestrator/popups/trigger_popup.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        result &= self.path_input.validate()
+
+        if self.trigger_type in (TriggerType.SINGLE, TriggerType.SCHEDULED):
+            result &= self.time_input.validate()
+
+            next_run = self.time_input.get_datetime()
+            if next_run and next_run < datetime.now():
+                result &= await generic_popups.question_popup(
+                        "The selected datetime is in the past. Do you want to create the trigger anyway?",
+                        "Create", "Cancel")
```
Comment (rimi-itk) [inline OpenOrchestrator/orchestrator/popups/trigger_popup.py:137]: Linguistically and pedantically, you cannot answer a yes/no question with “Create” or “Cancel”.

UI-wise best to have the “Create” and “Cancel” options, but then you should not ask a question – or change the reply options to “Yes, create the trigger” and “No, cancel”, say.

File: changelog.md [hunk 6]
```diff
@@ -7,6 +7,30 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- Orchestrator rewritten to use NiceGui instead of TKinter for its UI.
```
Comment (rimi-itk) [inline changelog.md:12]: Django coming up …

**Review by rimi-itk — CHANGES_REQUESTED** [review]
Looks good. Some comments added and some minor changes requested.

File: .github/workflows/Linting.yml (hunk 1 above)
Comment (ghbm-itk) [reply .github/workflows/Linting.yml:3]: I prefer it run on push so linting is run periodically.
This way litning is also run before a PR is even created

File: changelog.md (hunk 6 above)
Comment (ghbm-itk) [reply changelog.md:12]: image

File: OpenOrchestrator/common/datetime_util.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+from datetime import datetime
+
+
+def format_datetime(datetime_: datetime, default: str = 'N/A') -> str:
+    """Format a datetime to a string.
+
+    Args:
+        datetime_: The datetime to format.
+        default: A default string to return if the datetime is None. Defaults to 'N/A'.
```
Comment (rimi-itk) [inline OpenOrchestrator/common/datetime_util.py:12]: 🎉

**Pull request merged**